### ISSUE TITLE
Words not in CountVectorizer vocab despite being well above min_df threshold 

### ISSUE THREAD

**Issue by zilch42** [MaartenGr/BERTopic#1665]
Hi Maarten, 

I'm having an issue with some important words not appearing within CountVectorizer when using `min_df` even though they are well above the set threshold. My understanding of `min_df` is that when it is an integer (say, 10), all words that appear in at least 10 documents will be kept. Is that correct? See the example below,

```python
from umap import UMAP
from datasets import load_dataset
from bertopic import BERTopic
from sklearn.feature_extraction.text import CountVectorizer

dataset = load_dataset("CShorten/ML-ArXiv-Papers")["train"]
abstracts = dataset["abstract"][:5_000]

umap_model = UMAP(n_neighbors=15, n_components=5, min_dist=0.0, metric='cosine', random_state=42)
vectorizer_model=CountVectorizer(min_df=10)

topic_model = BERTopic(umap_model=umap_model, vectorizer_model=vectorizer_model).fit(abstracts)
```

Take the word 'cancer' for example. 

```python
# this under counts as it doesn't consider words next to punctuation, but wont consider words inside words (e.g. robot in robotics)
len([d for d in abstracts if " cancer " in d.lower()])
>>>  32
```

'cancer' appears in at least 32 documents (not counting occurrences up against punctuation so should be well above `min_df`.

```python
topic_model.vectorizer_model.vocabulary_["cancer"]
```
returns a key error. 

Also try words: *privacy, reward, quantum, fuzzy, hashing, planning, player, robot, embeddings*

If I remove the min_df param and just use `CountVectorizer()`, then these words are all present. 

I first came across this with the word 'rust' which was an important word in a large topic with ~70 document occurrences and `min_df=10`. It isn't prominent in the arxiv dataset though. 

Do you know what is happening here?

**Comment by MaartenGr** [comment]
Before going into what might causing this, it is worthwhile to instead check whether this is the case when performing the tokenization with the vectorizer model instead. That way, we can be 100% sure about the counts. Could you check the counts instead with [build_analyzer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html#sklearn.feature_extraction.text.CountVectorizer.build_analyzer) and the tokenizer to see whether that indeed does not match?

**Comment by zilch42** [comment]
Thanks for the tip. 

Is this what you mean?

```python
analyzer = vectorizer_model.build_analyzer()
len([d for d in abstracts if "cancer" in analyzer(d)])
>>> 43
```

**Comment by zilch42** [comment]
I thought about that, but your KeyBERT example doesn't use `min_df` at all. Every topic should have words in that example. 

I wonder if it is that there are docs that don't have any words that line up with the topic words? But there are more words cut out of the vocab when doing the filtering on topic-occurrences than there would be with document-occurrences, and I've never seen this error in the first (default) case. The error seems to result from supplying CountVectorizer with a `vocabulary`. 

Does that KeyBERT example run for you?

The issue seems to arise when selecting representative documents. If it is the result of docs that don't have any words that line up with the topic words, those docs are likely not representative documents. Could they just be skipped in the similarity calculation?

**Comment by MaartenGr** [comment]
Ah right, force of habit. I indeed meant that there might be a chance that some topics end up without words by restricting the vocabulary. It would be the same with setting `min_df` too high.

> Does that KeyBERT example run for you?

I haven't had the time but perhaps can check it out later this week. 

> The issue seems to arise when selecting representative documents. If it is the result of docs that don't have any words that line up with the topic words, those docs are likely not representative documents. Could they just be skipped in the similarity calculation?

It is not exactly the result of words not lining up but the calculation of a distance metric that is getting incorrect values. The main issue lies here:

```python
sim_matrix = cosine_similarity(ctfidf, c_tf_idf[index])
```

This means that there are values in either one or both matrices that result in infinite values being created. My guess is that the c-TF-IDF calculation might create strange values if there are few to no words in a topic.

**Comment by zilch42** [comment]
Ok, I've figured it out. The docs inside BERTopic get cleaned internally by `_preprocess_text()` before being tokenized, so by creating a vocabulary outside of BERTopic, even if it is created from the same document set, results in words in the vocabulary that do not appear in the cleaned docs. This results in a df with zero's here:

https://github.com/MaartenGr/BERTopic/blob/fc9a51aee18011afbbf4045f30460c2970527c49/bertopic/vectorizers/_ctfidf.py#L69-L70

which results in inf values when dividing by zero here:

https://github.com/MaartenGr/BERTopic/blob/fc9a51aee18011afbbf4045f30460c2970527c49/bertopic/vectorizers/_ctfidf.py#L79-L82

which causes problems for `cosine_similarity()`.

This can be solved by preprocessing the docs in the same way before creating the vocabulary:
```python
from sklearn.datasets import fetch_20newsgroups
from keybert import KeyBERT

# Prepare documents 
docs = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))['data']

def preprocess_text(documents: np.ndarray):
        """ Basic preprocessing of text

        Steps:
            * Replace \n and \t with whitespace
            * Only keep alpha-numerical characters
        """
        cleaned_documents = [doc.replace("\n", " ") for doc in documents]
        cleaned_documents = [doc.replace("\t", " ") for doc in cleaned_documents]
        cleaned_documents = [re.sub(r'[^A-Za-z0-9 ]+', '', doc) for doc in cleaned_documents]
        cleaned_documents = [doc if doc != "" else "emptydoc" for doc in cleaned_documents]
        return cleaned_documents

docs = preprocess_text(docs)

# Extract keywords
kw_model = KeyBERT()
keywords = kw_model.extract_keywords(docs)

# Create our vocabulary
vocabulary = [k[0] for keyword in keywords for k in keyword]
vocabulary = list(set(vocabulary))

from bertopic import BERTopic
from sklearn.feature_extraction.text import CountVectorizer

vectorizer_model= CountVectorizer(vocabulary=vocabulary)
topic_model = BERTopic(vectorizer_model=vectorizer_model, verbose=True)
topics, probs = topic_model.fit_transform(docs)
```

It would be nice to just use the internal function `_preprocess_text` so that I didn't have to keep track of any changes to that function, but we are creating the vocabulary before the BERTopic object so I'm not sure if there is a cleaner way to do this. 

I'm happy for this to be closed, but you may wish to update the KeyBERT example in Tips and Tricks so that it actually runs (and you may have a more elegant solution than I have above 😄 ).

**Comment by zilch42** [comment]
Sure, most docs in newsgroups have at least one example but try doc[8]. It has a few

```python
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import CountVectorizer
import numpy as np
import re

# Prepare documents 
docs = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))['data']

def preprocess_text(documents: np.ndarray):
        """ Basic preprocessing of text

        Steps:
            * Replace \n and \t with whitespace
            * Only keep alpha-numerical characters
        """
        cleaned_documents = [doc.replace("\n", " ") for doc in documents]
        cleaned_documents = [doc.replace("\t", " ") for doc in cleaned_documents]
        cleaned_documents = [re.sub(r'[^A-Za-z0-9 ]+', '', doc) for doc in cleaned_documents]
        cleaned_documents = [doc if doc != "" else "emptydoc" for doc in cleaned_documents]
        return cleaned_documents

clean_docs = preprocess_text(docs)

pre_vectorizer_model=CountVectorizer(stop_words="english")
pre_vectorizer_model.fit([docs[8]])
vocabulary = set(pre_vectorizer_model.vocabulary_.keys())

pre_vectorizer_model=CountVectorizer(stop_words="english")
pre_vectorizer_model.fit([clean_docs[8]])
clean_vocabulary = set(pre_vectorizer_model.vocabulary_.keys())

vocabulary.difference(clean_vocabulary)
```
```
{'bruin', 'didn', 'fuhr', 'sabre', 've'}

```

```
docs[8]
```
```
"\n\n\nYeah, it's the second one.  And I believe that price too.  I've been trying\nto get a good look at it on the Bruin-Sabre telecasts, and wow! does it ever\nlook good.  Whoever did that paint job knew what they were doing.  And given\nFuhr's play since he got it, I bet the Bruins are wishing he didn't have it:)\n"
```
It looks like the removal of punctuation causes them to split differently. 

If you compare the vocabularies from the whole dataset, a lot of the different words involve numbers too, but again, it may be down to punctuation removal.

```python
pre_vectorizer_model=CountVectorizer(stop_words="english")
pre_vectorizer_model.fit(docs)
vocabulary = set(pre_vectorizer_model.vocabulary_.keys())

pre_vectorizer_model=CountVectorizer(stop_words="english")
pre_vectorizer_model.fit(clean_docs)
clean_vocabulary = set(pre_vectorizer_model.vocabulary_.keys())

vocabulary.difference(clean_vocabulary)
```

```
{'8337',
 'mwhwc',
 'dob',
 'mgga2a',
 'd8yq',
 'eb1l',
 '7lbs',
 'lk8rlhzrlk8v',
 'iinsfdop93nn',
 'kelleyb',
 'wa4e',
 'wzr',
 'm66l4i',
 'e_xv3',
 'm87',
 'mcng',
 'qvdz',
 '77pt',
 'pmjpeg11',
 'vokabul',
 '0vy4b',
 'u40q',
 'w4k',
 'ldmy',
 'kd4fkw',
 'vsbd',
 '0669',
 '3mh',
 'acxx',
 '4wj',
 'obqp',
 'xbl4',
 'p6x',
 'jn2',
...
```

**Comment by zilch42** [comment]
I didn't necessarily see it as I bug for my use case. I like the preprocessing, and I wouldn't necessarily want things like "couldn't" being transformed to ["couldn", "t"] which I think is what CountVectorizer would do on its own. It makes sense that I would have to replicate any cleaning you are doing internally on the outside if I wanted to preprocess a vocabulary. 

Is the mismatch in preprocessing causing/likely to cause any other issues? Or can it just be left as is?

**Comment by MaartenGr** [comment]
It is quite an uncommon error that pops up and often resolved with some simple settings (for example reducing words with c-TF-IDF's parameters) or indeed accessing the internal preprocessing, so I would leave it for now. That said, thanks for sharing this! It definitely helps understand how people are using some of its functionality and the issues they run into.
